crates/biome_js_analyze/src/lint/nursery/no_magic_numbers.rs

```diff
@@ -0,0 +1,240 @@
+use biome_analyze::{
+    context::RuleContext, declare_lint_rule, Ast, FixKind, Rule, RuleDiagnostic, RuleSource,
+};
+use biome_console::markup;
+use biome_deserialize_macros::Deserializable;
+use biome_js_syntax::{
+    AnyJsExpression, JsAssignmentExpression, JsCallExpression, JsSyntaxKind, JsUnaryOperator,
+    JsVariableDeclarator,
+};
+use biome_rowan::{AstNode, SyntaxNode, TextRange};
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+const MAX_ARRAY_LENGTH: i64 = (2_i64.pow(32)) - 1;
+
+#[derive(
+    Debug, Default, Clone, Serialize, Deserialize, Deserializable, PartialEq, Eq, JsonSchema,
+)]
+pub struct NoMagicNumbersConfig {
+    detect_objects: bool,
+    enforce_const: bool,
+    ignore: Vec<i64>,
+    ignore_array_indexes: bool,
+    ignore_default_values: bool,
+    ignore_class_field_initial_values: bool,
+}
+
+declare_lint_rule! {
+    /// Disallow magic numbers
+    ///
+    /// This rule aims to make code more maintainable by ensuring that special numbers
+    /// are declared as constants with meaningful names.
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// let total = price * 1.23; // Magic number for tax rate
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// const TAX_RATE = 1.23;
+    /// let total = price * TAX_RATE;
+    /// ```
+    pub NoMagicNumbers {
+        version: "1.0.0",
+        name: "noMagicNumbers",
+        language: "js",
+        sources: &[RuleSource::Eslint("no-magic-numbers")],
+        recommended: false,
+        fix_kind: FixKind::None,
+    }
+}
+
+#[derive(Debug)]
+pub struct NumberContext {
+    range: TextRange,
+    raw: String,
+    is_const_violation: bool,
+}
+
+impl Rule for NoMagicNumbers {
+    type Query = Ast<AnyJsExpression>;
+    type State = NumberContext;
+    type Signals = Option<Self::State>;
+    type Options = NoMagicNumbersConfig;
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let node = ctx.query();
+        let config = ctx.options();
+
+        match node {
+            AnyJsExpression::AnyJsLiteralExpression(literal_expr) => {
+                if let Some(literal) = literal_expr.as_js_number_literal_expression() {
+                    let value_token = literal.value_token().ok()?;
+                    let value_text = value_token.text();
+                    let value = value_text.parse::<i64>().ok()?;
+                    if is_allowed_number(literal.syntax(), value, config) {
+                        return None;
+                    }
+
+                    let is_const_violation = check_const_violation(literal.syntax(), config);
+
+                    Some(NumberContext {
+                        range: literal.range(),
+                        raw: value_token.text().to_string(),
```

Comment:
Pass the `literal` node instead, and then extract the range and the name inside the `diagnostic` function. It will avoid allocating a `String`